00X-tracing.md

```diff
@@ -0,0 +1,579 @@
+   * function, it will be invoked to retrieve the value(s) only if tracing is enabled for the
+   * category. (The tracing system currently requires multi-value counters to have exactly two
+   * values.)
+   */
+  value?: number | { [name: string]: number } | () => (number | { [name: string]: number });
+
+  /**
+   * Optional arguments for "begin", "end", and "instant" events. If specified, this must be a
+   * either a dictionary of name-value pairs or a function that returns a dictionary. If a
+   * function, it will be invoked to retrieve the args only if tracing is enabled for the category.
+   * Not used for "count" events.
+   */
+  args?: { [name: string]: any } | () => { [name: string]: any };
+}
+
+interface Tracing {
+  /**
+   * Checks whether tracing is enabled for a category or for any categories in an array. Tracing
+   * is enabled for a category if either recording is enabled for the category or there are one
+   * or more listeners for the category.
+   *
+   * @param category Required tracing category name or array of category names.
+   * @returns True if tracing is enabled for the category or for any of the categories in the
+   *   array.
+   */
+  isEnabled(category: string | string[]): boolean;
+
+  /**
+   * Gets a list of tracing categories that are enabled for recording.
+   */
+  recordingCategories(): string[];
+
+  /**
+   * Enables or disables recording for a tracing category or categories. Tracing event recording
+   * is automatically started when one or more categories are enabled via this method or via the
+   * node command-line. Recording is automatically stopped when it is disabled for all
+   * categories.
+   * *
+   * @param category Required tracing category name or array of category names.
+   * @param enable Optional value specifying wiether recording will be enabled for the specified
+   *   category or categories. Defaults to true.
+   */
+  enableRecording(category: string | string[], enable?: boolean): void;
+
+  /**
+   * Emits a tracing event using an event object.
+   *
+   * @param e The event to be emitted. The event object must specify one or more categories.
+   * @returns Nonzero event id (either provided or generated) if the event was emitted, or null if
+   *   the event was not emitted because tracing was not enabled for any of the event categories.
+   */
+  emit(e: TracingEvent): number | null;
+
+  /**
+   * Emits a tracing event using an event object, with categories specified separately.
+   *
+   * @param category Required tracing category name or array of one or more category names for
+   *   the tracing event. Overrides any categories specified in the event object.
+   * @param e The event to be emitted.
+   * @returns Nonzero event id (either provided or generated) if the event was emitted, or null if
+   *   the event was not emitted because tracing was not enabled for any of the event categories.
+   */
+  emit(category: string | string[], e: TracingEvent): number | null;
```

Comment:
I think this is redundant, maybe we can try to avoid overloading as much as possible as it is harder to optimize, and this API should try to be as fast as possible.